qiskit/transpiler/passes/optimization/inverse_cancellation.py

```diff
@@ -91,9 +97,16 @@ def _run_on_self_inverse(self, dag: DAGCircuit, self_inverse_gates: List[Gate]):
         Returns:
             DAGCircuit: Transformed DAG.
         """
+        op_counts = dag.count_ops()
+        if not self.self_inverse_gate_names.intersection(op_counts):
+            return dag
         # Sets of gate runs by name, for instance: [{(H 0, H 0), (H 1, H 1)}, {(X 0, X 0}]
-        gate_runs_sets = [dag.collect_runs([gate.name]) for gate in self_inverse_gates]
-        for gate_runs in gate_runs_sets:
+        for gate in self.self_inverse_gates:
+            gate_name = gate.name
+            gate_count = op_counts.get(gate_name, 0)
+            if gate_count <= 1:
+                continue
+            gate_runs = dag.collect_runs([gate_name])
```

Comment:
Given we calculate an intersection with `op_counts` at the top of the function, it feels like we could skip the `continue` logic by iterating over the intersection rather than manually redoing it with retrieving the counts here - that'd also remove the early return, since the for loop would already serve that purpose.

Reply:
I'm not sure we should in this case. I was thinking about it, but iterating over the intersection directly means the loop is iterating over a set and the execution order is potentially non-deterministic. I think this would be problematic from a reproducibility PoV.

Reply:
We could always sort the set - it's rather unlikely to be large enough to matter.